File: src/foundation/tappaas-cicd/lib/ts/src/instance.ts
// instance.ts — instance vs module (ADR-026 D6).
//
// config/<instance>.json names an INSTANCE; the module it belongs to is named
// by its source directory (`.moduleSource`), whose basename is the module's name.
// The instance name equals the module name only by default (D6.4), and `vmname`
// is an instance name too — so neither is ever used to find the module.

import { basename, join } from "path";

/**
 * A deployed config's source directory: `.moduleSource`, or `.location` — its
 * name until #609, still read for a config migration 0006 has not reached.
 * "" when neither is a non-empty string (a site-style `location` object never
 * counts: that is a physical place, not a directory).
 */
export function moduleSourceOf(raw: Record<string, unknown> | null | undefined): string {
  for (const v of [raw?.moduleSource, raw?.location]) {
    if (typeof v === "string" && v !== "") return v;
  }
  return "";
}

/** The module whose source lives at <location>: the directory's basename (D6.3). */
export function moduleOfLocation(location: string): string {
  return basename(location.replace(/\/+$/, ""));
}

/** The release source JSON of the module at <location>: <location>/<module>.json. */
export function moduleSourceJson(location: string): string {
  return join(location, `${moduleOfLocation(location)}.json`);
}

File: src/foundation/tappaas-cicd/controller/opnsense-controller/src/opnsense_controller/instance.py
"""instance.py — a deployed config's module source directory (ADR-026 D6.2, #682).

The Python twin of ``lib/ts/src/instance.ts``. A deployed config records where its
module's code lives in ``moduleSource``; before #609 the field was called
``location``, and migration 0006 renames it. Every reader accepts both for one
stable cycle — a site restored from a backup, or one that has not yet taken 0006,
still carries the old name.

Reading only ``location`` is what #682 was: after 0006 the provider directory came
back empty, no auto-pinhole compiled, and `reconcile`'s prune then deleted every
established pinhole as an orphan.

A ``location`` that is not a string is never a path: `site.json` uses that name for
a physical place (an object), which ADR-026 D6.2 calls out as the trap.
"""

from __future__ import annotations


def module_source_of(data: dict | None) -> str:
    """The module's source directory: ``moduleSource``, else a string ``location``.

    "" when neither is a non-empty string.
    """
    if not isinstance(data, dict):
        return ""
    for key in ("moduleSource", "location"):
        value = data.get(key)
        if isinstance(value, str) and value:
            return value
    return ""
